CRM-19517 allow price fields that use disabled financial types to sho…

seamuslee001 opened the pull request:
…w up on admin screens so that administrators can edit them

monishdeb commented:
Discussed the patch with @seamuslee001 on MM and the PR needs some work, marking with WIP

monishdeb commented:
for detail https://issues.civicrm.org/jira/browse/CRM-19517?focusedCommentId=95519&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-95519

seamuslee001 commented:
@monishdeb I've fixed the public display issue and i have added upgrade. I am unsure how to handle the enable disable thing because it uses the API to enable or disable the option.

seamuslee001 commented:
@monishdeb I've chucked in exception now and i think this will do the trick if you can review that would be grand.

monishdeb commented:
@seamuslee001 I have tested your patch after creating price set with price field 'price A' and 'price B' (both Text/numeric) which are configured with FT A and FT B respectively. So I encountered few issues : 
1.  I was able to disable the 'FT A' which we didn't discussed in MM yet. But I think the correct behaivour would be to throw error message like ```Unable to disable the FT A as it is currently in use by one or more active price set(s).```

2. I was able to enable 'price A' which got disabled 'FT A' from UI. So to prevent this with desired message say ```Unable to enable 'price A' as Financial Type is either disabled or does not exist```

You can see the example of PriceSet (PS) where if you want to disable one PS which already been used in ContributionPage/Event/Membership from UI, then it throws a error message. You will find the code here 
https://github.com/civicrm/civicrm-core/blob/master/CRM/Admin/Page/AJAX.php#L95

seamuslee001 commented:
@monishdeb re 3 I don't think we can move it as we need the exception where it is right now for API purposes so i think we need to do a try and catch in that Line 163. 

I don't think we should necessarily stop someone from disabling a field here but at least give them a warning.  I think its a slight different kettle of fish to the whole price set

monishdeb commented on CRM/Utils/Check/Component/PriceFields.php:
```diff
@@ -0,0 +1,76 @@
+      INNER JOIN civicrm_price_field_value pfv ON pfv.price_field_id = psf.id
+      LEFT JOIN civicrm_financial_type cft ON cft.id = pfv.financial_type_id
+      WHERE cft.id IS NULL OR cft.is_active = 0";
+    $dao = CRM_Core_DAO::executeQuery($sql);
```
You can directly fetch the result in array by 
```$resultInArray = CRM_Core_DAO::executeQuery($sql)->toArray();``` recommended when result entries is < 100 .

monishdeb commented:
Agree with the final changes and now we have notification for price fields which use disabled FT like 
![image](https://cloud.githubusercontent.com/assets/3735621/19592704/87a92698-979a-11e6-9b63-b5c447f4fb33.png)

As discussed with @seamuslee001 on MM we decided to allow user to disable/enable FT and price field respectively as  stated in my last comment point 1 and 2. Since we already notifying user in ```System Check``` page. 

@eileenmcnaughton @JoeMurray any thoughts on this improvement ?

eileenmcnaughton commented on CRM/Price/BAO/PriceField.php:
```diff
@@ -91,16 +91,16 @@ public static function create(&$params) {
-
-    if ($priceField->html_type == 'Text') {
+    $priceField2 = civicrm_api3('price_field', 'get', array('id' => $priceField->id));
+    $priceField2 = $priceField2['values'][$priceField->id];
```
could use getsingle to lose one line

eileenmcnaughton commented on CRM/Price/BAO/PriceFieldValue.php:
```diff
@@ -125,6 +124,16 @@ public static function create(&$params, $ids = array()) {
         }
       }
     }
+    if ($id) {
```
Shouldn't passed in financial_type_id take precedence?

eileenmcnaughton commented on CRM/Price/BAO/PriceFieldValue.php:
```diff
@@ -125,6 +124,16 @@ public static function create(&$params, $ids = array()) {
+    else {
+      $financialType = CRM_Utils_Array::value('financial_type_id', $params, NULL);
+    }
+    CRM_Financial_BAO_FinancialType::getAvailableFinancialTypes($financialTypes);
```
There is actually a validateFinancialTypes function - perhaps this should go inside that?

seamuslee001 replied:
Can you point it at me?

seamuslee001 commented:
@eileenmcnaughton can we merge this if tests pass?

eileenmcnaughton commented:
I'm going to make sure it's assigned to someone in QA week because I haven't looked carefully enough & we have some JMA people signed up

JoeMurray commented on CRM/Utils/Check/Component/PriceFields.php:
```diff
@@ -0,0 +1,76 @@
+class CRM_Utils_Check_Component_PriceFields extends CRM_Utils_Check_Component {
+
+  /**
+   * Display warning about invlid priceFields
```
invalid

JoeMurray commented on CRM/Upgrade/Incremental/sql/4.7.14.mysql.tpl:
```diff
@@ -14,3 +14,9 @@ UPDATE civicrm_navigation SET
 WHERE `name` = 'Tags (Categories)'
 AND `url` = 'civicrm/admin/tag?reset=1';
+
+--  CRM-19517 Disable all price fields nad price field options that use disabled fianncial types
```
s/nad/and/
s/fianncial/financial/

JoeMurray commented on CRM/Upgrade/Incremental/sql/4.7.14.mysql.tpl:
```diff
@@ -14,3 +14,9 @@ UPDATE civicrm_navigation SET
+UPDATE civicrm_price_field_value cpfv
+INNER JOIN civicrm_financial_type cft oN cft.id = cpfv.financial_type_id
+SET cpfv.is_active = 0
+WHERE cft.is_active = 0;
```
So it looks like this disables price field option values if its FT is disabled. Can you also disable price fields with no options if their FTs are disabled? Also, elsewhere in PR you handle case of null FTs (!); if you have seen evidence of them please handle setting is_active = false as appropriate using a left join to include those records as above.

seamuslee001 replied:
@JoeMurray  should have a solution for that shortly, and with your other comment if your meaning this https://github.com/civicrm/civicrm-core/pull/9276/files#diff-2bd325d7967360f15458894201f5cba9R128 then all that is doing is saying are we getting a financial type passed through otherwise load it from the current record (in case someone is just clicking enable where its doing an API call in without passing in any new financial type info)

JoeMurray commented on CRM/Price/Page/Field.php:
```diff
@@ -142,12 +142,8 @@ public function browse() {
         CRM_Price_BAO_PriceFieldValue::retrieve($params, $optionValues);
-        $financialTypeId = $optionValues['financial_type_id'];
-        if (!array_key_exists($financialTypeId, $financialTypes)) {
-          unset($priceField[$priceFieldBAO->id]);
```
@Edzelopez we must have miscommunicated about this somehow. :(

seamuslee001 commented:
@JoeMurray Thanks for that pick up re the options. I think i got that fixed now and also fixed the upgrade

seamuslee001 commented:
Jenkins test this please

eileenmcnaughton commented:
@JoeMurray are you happy that this is ready to merge now? @seamuslee001 should we squash into 1 commit?

JoeMurray commented:
Jenkins test this please

pradpnayak commented:
Can someone rebase the PR?

monishdeb commented:
@pradpnayak PR rebased

pradpnayak commented:
Thanks @monishdeb I will start with QA now. I will post result of QA soon.

seamuslee001 commented:
@JoeMurray @pradpnayak have rebased again can you please QA this and hopefully we can get this in for 4.7.16

seamuslee001 commented:
@JoeMurray @pradpnayak Can one of you QA this so we can get this into the next release?

jitendrapurohit commented:
I'll QA this and comment as soon as its done.

jitendrapurohit commented on CRM/Price/Form/Preview.php:
```diff
@@ -132,7 +132,9 @@ public function buildQuickForm() {
-          CRM_Price_BAO_PriceField::addQuickFormElement($this, $elementName, $fieldId, FALSE, $field['is_required']);
+          if (!empty($field['options'])) {
+            CRM_Price_BAO_PriceField::addQuickFormElement($this, $elementName, $fieldId, FALSE, $field['is_required']);
+          }
```
Due to this `Preview` action of a text field displays nothing. To replicate - Create a text price field and click on Preview. As a text field doesn't have any options, no element is being added.

seamuslee001 replied:
I assume this happens with the other if statements right?

jitendrapurohit replied:
could be, I just found this on initial QA. Also, not sure why the merge build isn't completing.

seamuslee001 replied:
@jitendrapurohit i just re-tested on a locally re-built buildkit and couldn't reproduce the issue  as per attached image 
![price_text_field](https://cloud.githubusercontent.com/assets/6799125/24681857/4b1be982-19da-11e7-9613-07c6d71d4bfa.jpg)

jitendrapurohit replied:
This is I think when you preview the whole set, I was mentioning about the field (i.e) when you click on preview against the text custom field.

seamuslee001 replied:
I am sure i checked both this morning and it displayed as i expected

jitendrapurohit replied:
OK, so this is when FT of the price field is disabled. You can see in the field list, but Preview link shows empty.

<img width="695" alt="image" src="https://cloud.githubusercontent.com/assets/5929648/24690335/ebfef25e-19e9-11e7-80cb-9531ce6c6bf2.png">

Preview -

<img width="507" alt="image" src="https://cloud.githubusercontent.com/assets/5929648/24690346/fd1504c0-19e9-11e7-8608-467f258c0dee.png">

Not sure if this is the expected behavior?

seamuslee001 replied:
I would have thought so as we don't want fields with disabled FTs showing publicly but we want them able to be accessible on the backend so admins can alter the FT

eileenmcnaughton replied:
That seems reasonable - shouldn't see something valid for an invalid config - but might be nice to add a session message

seamuslee001 commented:
Jenkins test this please

eileenmcnaughton commented:
@seamuslee001 you should squash those commits

jitendrapurohit commented on CRM/Price/BAO/PriceField.php:
```diff
@@ -94,22 +94,15 @@ public static function create(&$params) {
+      $fieldOptions = civicrm_api3('price_field_value', 'get', array(
+        'price_field_id' => $priceField->id,
+        'sequential' => 1,
+      ));
```
There might be some BAO function which would be doing this, shouldn't we avoid using apis under the BAO functions?

eileenmcnaughton replied:
no, we only avoid using api from the BAOs where it would be circular - ie. calling create from  create. In general the api functions are the most tested functions & are encouraged for most scenarios

jitendrapurohit commented on CRM/Price/BAO/PriceField.php:
```diff
@@ -94,22 +94,15 @@ public static function create(&$params) {
-        if (!empty($params['id']) && $priceField->label != $fieldValue->label) {
-          $fieldValue->label = $priceField->label;
-          $fieldValue->save();
-        }
```
Removal of this part seems to be causing a minor regression of https://issues.civicrm.org/jira/browse/CRM-19741. Should have written a test :(. Is this necessary to be removed?

eileenmcnaughton commented:
After discussing with @seamuslee001 Im going to merge this, on the basis that Seamus will address the minor regression @jitendrapurohit identified as a follow up fix. I think it would be good to get the main part of this merged given the work & time already in it.

@seamuslee001 please log a JIRA for the follow up fix (ahem did someone mention including a unit test in it...)
